commit 920fc941a9617f95ccb283037fe6f8a38d95bb69
Author: Thomas Graf <tgraf@suug.ch>
Date:   Thu Mar 27 16:08:03 2008 -0700

    [ESP]: Ensure IV is in linear part of the skb to avoid BUG() due to OOB access
    
    ESP does not account for the IV size when calling pskb_may_pull() to
    ensure everything it accesses directly is within the linear part of a
    potential fragment. This results in a BUG() being triggered when the
    both the IPv4 and IPv6 ESP stack is fed with an skb where the first
    fragment ends between the end of the esp header and the end of the IV.
    
    This bug was found by Dirk Nehring <dnehring@gmx.net> .
    
    Signed-off-by: Thomas Graf <tgraf@suug.ch>
    Signed-off-by: David S. Miller <davem@davemloft.net>


Adjusted to apply to Debian's 2.6.24 by dann frazier <dannf@debian.org>

diff -urpN linux-source-2.6.24.orig/net/ipv4/esp4.c linux-source-2.6.24/net/ipv4/esp4.c
--- linux-source-2.6.24.orig/net/ipv4/esp4.c	2008-01-24 15:58:37.000000000 -0700
+++ linux-source-2.6.24/net/ipv4/esp4.c	2008-07-18 16:51:16.000000000 -0600
@@ -165,7 +165,7 @@ static int esp_input(struct xfrm_state *
 	int padlen;
 	int err;
 
-	if (!pskb_may_pull(skb, sizeof(*esph)))
+	if (!pskb_may_pull(skb, sizeof(*esph) + esp->conf.ivlen))
 		goto out;
 
 	if (elen <= 0 || (elen & (blksize-1)))
diff -urpN linux-source-2.6.24.orig/net/ipv6/esp6.c linux-source-2.6.24/net/ipv6/esp6.c
--- linux-source-2.6.24.orig/net/ipv6/esp6.c	2008-01-24 15:58:37.000000000 -0700
+++ linux-source-2.6.24/net/ipv6/esp6.c	2008-07-18 16:51:16.000000000 -0600
@@ -155,7 +155,7 @@ static int esp6_input(struct xfrm_state 
 	int nfrags;
 	int ret = 0;
 
-	if (!pskb_may_pull(skb, sizeof(*esph))) {
+	if (!pskb_may_pull(skb, sizeof(*esph) + esp->conf.ivlen)) {
 		ret = -EINVAL;
 		goto out;
 	}
